### PR TITLE
Distillation Algorithm

### PR DESCRIPTION
**Pull request by growlix**
# What does this PR do?

This PR adds a distillation algorithm to Composer. We found that distillation is a [consistent efficiency improvement for vision and MLM models](https://arxiv.org/abs/2211.00683). This implementation supports multiple teachers, but we saved loading checkpoints directly from cloud storage for the future.

@codestar12: It's mostly your code.
@bcui19: ~Distillation loads + performs inference on a second model in parallel with the model being trained. Do you have any suggestions on how to make this compatible with FSDP?~ Thanks for the help w/ FSDP compatibility. I also implemented what I think is the appropriate DDP compatibility, but I'm not sure it's necessary given that the teacher doesn't receive gradients.
@Landanjs: Because distillation.

# Before submitting
- [ ] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [ ] Is this change a documentation change or typo fix? If so, skip the rest of this checklist.
- [ ] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [ ] Did you update any related docs and document your change?
- [ ] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [ ] Did you run the tests locally to make sure they pass?
- [ ] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

### CODE REVIEW

**Comment by mvpatel2000** [comment]
> @mvpatel2000: Distillation loads + performs inference on a second model in parallel with the model being trained. Do you have any suggestions on how to make this compatible with FSDP?
It should be fine as long as the model class is wrapped properly for FSDP -- eg both models are sharded so that they fit together. @bcui19 can help here

File: composer/algorithms/distillation/distillation.py [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+        softmax_dim: int = -1,
+    ):
+        super(KLDivergence, self).__init__()
+        self.temperature = temperature
+        self.softmax_dim = softmax_dim
+
+    def forward(self, y_student: torch.Tensor, y_teacher: torch.Tensor):
+        p_student = F.log_softmax(y_student / self.temperature, dim=self.softmax_dim)
+        p_teacher = F.softmax(y_teacher / self.T, dim=self.softmax_dim)
+        loss = F.kl_div(p_student, p_teacher, size_average=False) * (self.temperature**2) / y_student.shape[0]
```
Comment (codestar12) [inline composer/algorithms/distillation/distillation.py:192]: I know this is my code, but I think size_average is being deprecated in the near future

I think 

```suggestion
        loss = F.kl_div(p_student, p_teacher, reduction='batchmean') * (self.temperature**2) 
```

should be equivalent.

**Comment by nik-mosaic** [comment]
Missing README.md and metadata.json?

File: composer/algorithms/distillation/README.md [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+KnowledgeDistillation
+The process of knowledge distillation. The teacher model makes predictions for a set of inputs, and the student model is trained to mimic these predictions. This image is from Chen et al., 2015.
+How to Use
+To use knowledge distillation, you will need to have both a teacher model and a student model. You can then train the student model by providing it with the predictions of the teacher model as "soft targets" (i.e., predicted probabilities for each class), in addition to the ground truth labels. The student model can be trained using standard supervised learning techniques, such as cross-entropy loss.
```
Comment (bcui19) [inline composer/algorithms/distillation/README.md:11]: Looks like this is extraneous/duplicated from below and above?

File: composer/algorithms/distillation/README.md [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+
+## How to use
+
+To use knowledge distillation, you will need to have both a teacher model and a student model. You can then train the student model by providing it with the predictions of the teacher model as "soft targets" (i.e., predicted probabilities for each class), in addition to the ground truth labels. The student model can be trained using standard supervised learning techniques, such as cross-entropy loss.
+
+### Composer Trainer With Composer Checkpoints
```
Comment (bcui19) [inline composer/algorithms/distillation/README.md:17]: Maybe rename the subheader to sth like: 'Composer Trainer with Composer Models' 

just so it doesn't match the below subsubheader

File: composer/algorithms/distillation/README.md [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+)
+
+trainer.fit()
+```
+
+
+
+### Implementation Details
+
+The Distillation works by calculating the loss between the teacher and student models after the student model completes its forward pass and modifying the resulting loss in the trainer. `org_loss_weight` and `kd_loss_weight` can be used to balance loss of the ground truth training objective and the knowledge distillation loss. This particular implementation
```
Comment (bcui19) [inline composer/algorithms/distillation/README.md:123]: TODO: fill in missing details on implementation?

File: composer/algorithms/distillation/distillation.py [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+        softmax_dim: int = -1,
+    ):
+        super(KLDivergence, self).__init__()
+        self.temperature = temperature
+        self.softmax_dim = softmax_dim
+
+    def forward(self, y_student: torch.Tensor, y_teacher: torch.Tensor):
+        p_student = F.log_softmax(y_student / self.temperature, dim=self.softmax_dim)
+        p_teacher = F.softmax(y_teacher / self.temperature, dim=self.softmax_dim)
+        loss = F.kl_div(p_student, p_teacher, reduction='batchmean') * (self.temperature**2)
```
Comment (bcui19) [inline composer/algorithms/distillation/distillation.py:249]: I haven't tested this, but have we run into numerical issues w/ the target ('teacher') not being in log space:

https://pytorch.org/docs/stable/generated/torch.nn.functional.kl_div.html

File: composer/algorithms/distillation/README.md [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+
+
+## Suggested Hyperparameters
+There are a few hyperparameters that you may want to tune when using knowledge distillation:
+
+temperature: This hyperparameter controls the "sharpness" of the soft targets provided by the teacher model. A higher temperature will result in softer targets, while a lower temperature will result in harder targets. In general, a temperature of 1 works well, but you may want to try higher or lower values depending on your specific setup.
+
+alpha: This hyperparameter controls the balance between the distillation loss and the ground truth loss. A value of 1 will give equal weight to both losses, while a value less than 1 will give more weight to the distillation loss and a value greater than 1 will give more weight to the ground truth loss.
```
Comment (bcui19) [inline composer/algorithms/distillation/README.md:131]: Is this the `kd_loss_weight`?

File: composer/algorithms/distillation/README.md [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+Use the Kullback-Leibler divergence loss when training the student model with the soft targets. This loss measures the difference between two probability distributions, making it well-suited for comparing the predictions of the student model with the soft targets.
+
+The teacher model should be set to eval mode and have gradients disabled during training to save computation and memory.
+
+The student model should be set to train mode and have gradients enabled during training to allow for backpropagation of the loss.
+
+## Attribution
+Knowledge distillation was first introduced in the following paper:
+
+Hinton,
```
Comment (bcui19) [inline composer/algorithms/distillation/README.md:147]: TODO: cite

File: composer/algorithms/distillation/distillation.py [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+            with torch.no_grad():
+                with get_precision_context(state.precision):
+                    for idx in t_idx:
+                        teacher_output = self.teachers[int(idx)](input)
+                        if isinstance(teacher_output, MaskedLMOutput):
+                            teacher_outputs.append(teacher_output.logits.detach().clone())
+                        else:
+                            teacher_outputs.append(teacher_output)
+            # Avg teacher outputs
+            teacher_output = torch.mean(torch.stack(teacher_outputs), dim=0)
```
Comment (bcui19) [inline composer/algorithms/distillation/distillation.py:171]: Is this the correct/desired formulation? 

E.g. is the KL_Div(student || avg_teacher) == avg(KL_DIV(student || teacher_i))? 

I don't think it is, but I could be wrong?

Comment (bcui19) [reply composer/algorithms/distillation/distillation.py:171]: But also trying to match a single distribution might be easier than matching 'N' distributions empirically -- at least intuitively. I don't know what's common best practice or if there is one though

Comment (codestar12) [reply composer/algorithms/distillation/distillation.py:171]: This is the desired formulation. I'm not sure if matching the individual distributions is equivalent, but I do know that calculating N teacher losses is going to be slower. In our experiments we have mostly been using MSE for multi teacher distillation anyways so it really doesn't matter.

File: composer/algorithms/distillation/README.md (hunk 2 above)
Comment (codestar12) [reply composer/algorithms/distillation/README.md:11]: Most of the readme needs to be re-written. I was fighting getting it to pass tests and gave up just before christmas.

**Comment by mvpatel2000** [comment]
Closing for now
